cropduster/static/cropduster/js/cropduster.js

```diff
@@ -372,6 +372,18 @@ window.CropDuster = {};
                     $thumb.html($thumb.html() + $.render.cropdusterImage(thumbData[size.name]));
                 }
             });
+            var url = [CropDuster.mediaUrl, path, '_preview' + ext].join('/');
```

Comment:
The code in this file is a nightmare, so it's an easy mistake but: I think this introduces a bug. This uses the definition of `url` in the loop above. But notice that we don't actually pass that value to the template renderer any more. If the media storage is something other than django's FileSystemStorage, we can't reliably construct the url in this way; we have to get it from the server.

Reply:
Confirmed. This assertion is removed in this PR, but if put back in the S3 test fails:

```patch
diff --git a/cropduster/tests/test_admin.py b/cropduster/tests/test_admin.py
index 7d7a8ed..1b3c158 100644
--- a/cropduster/tests/test_admin.py
+++ b/cropduster/tests/test_admin.py
@@ -148,7 +148,6 @@ def test_changeform_single_image(self):
 
         preview_image_el = self.selenium.find_element_by_css_selector('#headshot-group .cropduster-image span')
         src_image_path = os.path.join(self.TEST_IMG_DIR, 'img.png')
-        self.assertImageColorEqual(preview_image_el, src_image_path)
 
         elem = self.selenium.find_element_by_id('id_name')
         elem.clear()
```

Reply:
You'll need to add a data attribute here with the preview image's url:

https://github.com/theatlantic/django-cropduster/blob/bc251f421ea2b8e8a7370ad55d221394db65e344/cropduster/templates/cropduster/custom_field.html#L6-L12

You can pass context data to that template here:

https://github.com/theatlantic/django-cropduster/blob/bc251f421ea2b8e8a7370ad55d221394db65e344/cropduster/forms.py#L31-L46

Reply:
A-ha, very helpful. Cheers :+1: I'll give that a go.

Reply:
That will ensure that the preview image shows up when the admin changeview is loaded. In order to also update that image after a user has finished uploading and cropping an image, you'll need to add the preview url to the crop view's response:

https://github.com/theatlantic/django-cropduster/blob/178cb00a6a19660d40235e8ec30c65175c158f62/cropduster/views/__init__.py#L468-L472

which gets passed along in upload.js to `parent.CropDuster.complete`:

https://github.com/theatlantic/django-cropduster/blob/178cb00a6a19660d40235e8ec30c65175c158f62/cropduster/static/cropduster/js/upload.js#L703

which is defined here:

https://github.com/theatlantic/django-cropduster/blob/178cb00a6a19660d40235e8ec30c65175c158f62/cropduster/static/cropduster/js/cropduster.js#L235-L250

Where you'll need to do something like this before the call to `createThumbnails`:

```javascript
$("#id_" + prefix).data('previewUrl', data.preview_url);
```

Reply:
Done 👍